Content.Server/MassDriver/EntitySystems/MassDriverSystem.cs

```diff
@@ -0,0 +1,336 @@
+    private void OnPortDisconnected(EntityUid uid, MassDriverConsoleComponent component, PortDisconnectedEvent args)
+    {
+        var massDriverNetEntity = GetNetEntity(args.Sink);
+        if (args.Port != component.LinkingPort || !component.MassDrivers.Contains(massDriverNetEntity))
+            return;
+
+        var massDriverEntityUid = args.Sink;
+        if (TryComp<MassDriverComponent>(massDriverEntityUid, out var massDriverComponent))
+        {
+            massDriverComponent.Console = null;
+            Dirty(massDriverEntityUid, massDriverComponent);
+        }
+
+        component.MassDrivers.Remove(massDriverNetEntity);
+        Dirty(uid, component);
+    }
+
+    /// <summary>
+    /// Handle signal receive, for launch some objects.
+    /// </summary>
+    /// <param name="uid">Mass Driver</param>
+    /// <param name="component">Mass Driver Component</param>
+    /// <param name="args">Event arguments</param>
+    private void OnSignalReceived(EntityUid uid, MassDriverComponent component, ref SignalReceivedEvent args)
+    {
+        if (args.Port == component.LaunchPort && component.Mode == MassDriverMode.Manual)
+            AddComp<ActiveMassDriverComponent>(uid);
+    }
+
+    #endregion
+
+    #region Logic
+
+    /// <summary>
+    /// Handle power changing to disable or enable mass driver, so it can't work without power
+    /// </summary>
+    /// <param name="uid">Mass Driver</param>
+    /// <param name="component">Mass Driver Component</param>
+    /// <param name="args">Event arguments</param>
+    private void OnPowerChanged(EntityUid uid, MassDriverComponent component, ref PowerChangedEvent args)
+    {
+        if (component.Mode != MassDriverMode.Auto)
+            return;
+
+        var hasComp = HasComp<ActiveMassDriverComponent>(uid);
+        if (hasComp && !args.Powered)
+            RemComp<ActiveMassDriverComponent>(uid);
+        else if (!hasComp && args.Powered)
+            EnsureComp<ActiveMassDriverComponent>(uid);
+    }
+
+    /// <summary>
+    /// Update only active mass drivers, so we have it more optimized than conveyors
+    /// </summary>
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<ActiveMassDriverComponent, MassDriverComponent, ApcPowerReceiverComponent>();
+        while (query.MoveNext(out var uid, out var activeMassDriver, out var massDriver, out var powered))
+        {
+            if (_timing.CurTime < activeMassDriver.NextUpdateTime)
+                continue;
+
+            if (activeMassDriver.NextUpdateTime == TimeSpan.Zero)
+            {
+                activeMassDriver.NextUpdateTime = _timing.CurTime + activeMassDriver.UpdateDelay;
+                continue;
+            }
+            activeMassDriver.NextUpdateTime = _timing.CurTime + activeMassDriver.UpdateDelay;
+
+            if (!_powerReceiver.IsPowered(uid))
+                continue;
+
```

Comment:
This should just be a debug assert as it's already being adjusted if the powerchangedevent runs anyway.